# Housing Prices in San Francisco (Craigslist) - Regression with PyCaret
## Predicting housing prices based on scraped craigslist data
This version was made after the Kaggle Version without using PyCaret and was made in colab.

In [4]:
#pip install pycaret

In [3]:
# import libraries
import pandas as pd
import numpy as np
from pycaret.regression import *

In [5]:
# load data
df=pd.read_csv("sf_clean.csv")
df.head()

price    sqft  beds  bath         laundry         pets housing_type  \
0   6800  1600.0   2.0   2.0     (a) in-unit  (d) no pets    (c) multi   
1   3500   550.0   1.0   1.0     (a) in-unit     (a) both    (c) multi   
2   5100  1300.0   2.0   1.0     (a) in-unit     (a) both    (c) multi   
3   9000  3500.0   3.0   2.5     (a) in-unit  (d) no pets    (c) multi   
4   3100   561.0   1.0   1.0  (c) no laundry     (a) both    (c) multi   

          parking  hood_district  
0   (b) protected            7.0  
1   (b) protected            7.0  
2  (d) no parking            7.0  
3   (b) protected            7.0  
4  (d) no parking            7.0

In [6]:
setup(data=df,target="price") #instead of doing everything manually, this code will do the work

Description             Value
0                               session_id              7053
1                                   Target             price
2                            Original Data          (989, 9)
3                           Missing Values             False
4                         Numeric Features                 4
5                     Categorical Features                 4
6                         Ordinal Features             False
7                High Cardinality Features             False
8                  High Cardinality Method              None
9                    Transformed Train Set         (692, 18)
10                    Transformed Test Set         (297, 18)
11                      Shuffle Train-Test              True
12                     Stratify Train-Test             False
13                          Fold Generator             KFold
14                             Fold Number                10
15                                CPU Jobs                -1
16                                 Use GPU             False
17                          Log Experiment             False
18                         Experiment Name  reg-default-name
19                                     USI              f339
20                         Imputation Type            simple
21          Iterative Imputation Iteration              None
22                         Numeric Imputer              mean
23      Iterative Imputation Numeric Model              None
24                     Categorical Imputer          constant
25  Iterative Imputation Categorical Model              None
26           Unknown Categoricals Handling    least_frequent
27                               Normalize             False
28                        Normalize Method              None
29                          Transformation             False
30                   Transformation Method              None
31                                     PCA             False
32                              PCA Method              None
33                          PCA Components              None
34                     Ignore Low Variance             False
35                     Combine Rare Levels             False
36                    Rare Level Threshold              None
37                         Numeric Binning             False
38                         Remove Outliers             False
39                      Outliers Threshold              None
40                Remove Multicollinearity             False
41             Multicollinearity Threshold              None
42             Remove Perfect Collinearity              True
43                              Clustering             False
44                    Clustering Iteration              None
45                     Polynomial Features             False
46                       Polynomial Degree              None
47                    Trignometry Features             False
48                    Polynomial Threshold              None
49                          Group Features             False
50                       Feature Selection             False
51                Feature Selection Method           classic
52            Features Selection Threshold              None
53                     Feature Interaction             False
54                           Feature Ratio             False
55                   Interaction Threshold              None
56                        Transform Target             False
57                 Transform Target Method           box-cox

INFO:logs:create_model_container: 0
INFO:logs:master_model_container: 0
INFO:logs:display_container: 1
INFO:logs:Pipeline(memory=None,
         steps=[('dtypes',
                 DataTypes_Auto_infer(categorical_features=[],
                                      display_types=True, features_todrop=[],
                                      id_columns=[], ml_usecase='regression',
                                      numerical_features=[], target='price',
                                      time_features=[])),
                ('imputer',
                 Simple_Imputer(categorical_strategy='not_available',
                                fill_value_categorical=None,
                                fill_value_numerical=None,
                                numeric_strategy='...
                ('scaling', 'passthrough'), ('P_transform', 'passthrough'),
                ('binn', 'passthrough'), ('rem_outliers', 'passthrough'),
                ('cluster_all', 'passthrough'),
              

(KFold(n_splits=10, random_state=None, shuffle=False), 629    5100
 985    4695
 768    3104
 568    3775
 50     2600
        ... 
 908    3395
 16     3700
 56      750
 907    2495
 529    2511
 Name: price, Length: 692, dtype: int64, False, {'USI',
  'X',
  'X_test',
  'X_train',
  '_all_metrics',
  '_all_models',
  '_all_models_internal',
  '_available_plots',
  '_gpu_n_jobs_param',
  '_internal_pipeline',
  '_ml_usecase',
  'create_model_container',
  'dashboard_logger',
  'data_before_preprocess',
  'display_container',
  'exp_name_log',
  'experiment__',
  'fix_imbalance_method_param',
  'fix_imbalance_param',
  'fold_generator',
  'fold_groups_param',
  'fold_groups_param_full',
  'fold_param',
  'fold_shuffle_param',
  'gpu_param',
  'html_param',
  'imputation_classifier',
  'imputation_regressor',
  'iterative_imputation_iters_param',
  'log_plots_param',
  'logging_param',
  'master_model_container',
  'n_jobs_param',
  'prep_pipe',
  'pycaret_globals',
  'seed',
  'strati

In [7]:
best_model=compare_models() # see the results

Model        MAE           MSE       RMSE  \
llar         Lasso Least Angle Regression   501.7014  5.572094e+05   737.9921   
lasso                    Lasso Regression   508.8053  5.634815e+05   742.3662   
br                         Bayesian Ridge   506.8143  5.642778e+05   742.9392   
ridge                    Ridge Regression   509.9901  5.652721e+05   743.4987   
lr                      Linear Regression   510.5401  5.659979e+05   743.9821   
lar                Least Angle Regression   510.5397  5.659966e+05   743.9812   
huber                     Huber Regressor   496.5327  5.754592e+05   747.6928   
en                            Elastic Net   516.0784  5.768920e+05   752.6167   
rf                Random Forest Regressor   504.0494  6.142645e+05   768.5802   
lightgbm  Light Gradient Boosting Machine   508.7785  6.505690e+05   781.9848   
omp           Orthogonal Matching Pursuit   541.0919  6.078737e+05   773.7400   
et                  Extra Trees Regressor   528.1349  6.169541e+05   775.7220   
gbr           Gradient Boosting Regressor   506.6688  6.666035e+05   797.5247   
knn                 K Neighbors Regressor   588.6825  6.965222e+05   827.6199   
ada                    AdaBoost Regressor   611.1775  8.419124e+05   888.6974   
dt                Decision Tree Regressor   654.8128  1.119605e+06  1040.0890   
par          Passive Aggressive Regressor  1031.0805  2.121036e+06  1359.4383   
dummy                     Dummy Regressor  1115.7412  2.454362e+06  1549.0469   

              R2   RMSLE    MAPE  TT (Sec)  
llar      0.7658  0.1717  0.1373     0.013  
lasso     0.7628  0.1736  0.1391     0.014  
br        0.7623  0.1728  0.1384     0.015  
ridge     0.7620  0.1741  0.1396     0.012  
lr        0.7617  0.1743  0.1397     0.348  
lar       0.7617  0.1743  0.1397     0.015  
huber     0.7612  0.1715  0.1345     0.048  
en        0.7548  0.1752  0.1410     0.013  
rf        0.7452  0.1702  0.1348     0.302  
lightgbm  0.7419  0.1716  0.1347     0.115  
omp       0.7398  0.1833  0.1491     0.012  
et        0.7382  0.1774  0.1423     0.268  
gbr       0.7236  0.1695  0.1332     0.078  
knn       0.6968  0.1956  0.1604     0.015  
ada       0.6526  0.2096  0.1803     0.090  
dt        0.5149  0.2234  0.1727     0.018  
par       0.0608  0.3049  0.2854     0.016  
dummy    -0.0160  0.3958  0.3500     0.010

INFO:logs:create_model_container: 18
INFO:logs:master_model_container: 18
INFO:logs:display_container: 2
INFO:logs:LassoLars(alpha=1.0, copy_X=True, eps=2.220446049250313e-16, fit_intercept=True,
          fit_path=True, jitter=None, max_iter=500, normalize=True,
          positive=False, precompute='auto', random_state=7053, verbose=False)
INFO:logs:compare_models() succesfully completed......................................


Using PyCaret we achieved lower results than our model without PyCaret. Here our best result is with **Lasso Least Angle Regression with R2 of 73,58%, MAE of 501.7 and RMSE of 737,99**